### ISSUE TITLE
OIDC-login redirect Location header points to REMS instead of the IdP

### ISSUE THREAD

**Issue by katpavlova** [CSCfi/rems#2510]
As a developer, I would like to have REMS authenticate users with a local Keycloak IdP. From the REMS homepage, the user should click `Login` and be redirected to Keycloak for authentication. Upon authentication, the user should be redirected back to REMS.

For the description below, assume that my local REMS instance is exposed at `http://127.0.0.1:3001`, and that my local Keycloak instance is exposed at `https://8556b3fbb6fb.ngrok.io`.

In the past (ex. December 7th commit `2f8c02a0ac68e2da2fb6cc477535b8e791760f01`), I would set `:oidc-domain` parameter to the `host:port/realm` pertaining to the relevant Keycloak realm: `8556b3fbb6fb.ngrok.io/auth/realms/dycons-researcher-idp`. When I navigated my browser to my local REMS instance and clicked the `Login` button, I would be redirected to Keycloak for authentication, then redirected back to REMS following authentication. See the picture below for an example of the `302 Found` response that REMS would use to redirect my browser after `Login` was clicked: 
![rems-oidc-02](https://user-images.githubusercontent.com/19292858/103703243-2f675080-4f75-11eb-87bc-21b67157a297.png)
Notice that the `Location` response header redirects to a URL that matches the `:oidc-domain` string.

Now (commit `648361b7042ad0910045898b0a0a385dff48f52a`), I set the `oidc-metadata-url` parameter to the full url of my Keycloak realm (`https://host:port/realm`): `https://8556b3fbb6fb.ngrok.io/auth/realms/dycons-researcher-idp`. When I navigate my browser to my local REMS instance and click the `Login` button, I am redirected to REMS instead of Keycloak. Therefore I am never actually redirected to Keycloak, and am unable to login. See the picture below for the `302 Found` response generated after `Login` is clicked:
![rems-oidc-01](https://user-images.githubusercontent.com/19292858/103703653-eebc0700-4f75-11eb-8729-93c2a3a60cd6.png)
Notice that the `Location` response header redirects to a URL whose host matches my REMS host.

It looks like the `Location` url generated for redirection by `oidc-login` might be using the REMS url instead of the url provided in the `oidc-metadata-url` parameter.

**Comment by Macroz** [comment]
I'll investigate more tomorrow when holiday is over but the `:oidc-metadata-url` should point to the configuration URL and not the login URL. Can you show me what the config JSON for Keycloak looks like?

**Comment by opqdonut** [comment]
If you previously used the configuration
```
:oidc-doman "8556b3fbb6fb.ngrok.io/auth/realms/dycons-researcher-idp"
```
you should now use
```
:oidc-metadata-url "https://8556b3fbb6fb.ngrok.io/auth/realms/dycons-researcher-idp/.well-known/openid-configuration"
```

**Comment by opqdonut** [comment]
We should probably improve error handling for oidc configuration...

**Comment by katpavlova** [comment]
Thank you very much, modifying the parameter like @opqdonut suggested worked.

By no means urgent, but it would probably be helpful to modify the error handling so that the user isn't redirected to the REMS instance if `:oidc-metadata-url` is misconfigured. Took me a while to notice that there was a redirection issue, because it just looked like the page was being refreshed every time I hit "Login."

I'm sorry it took me a couple days to respond, after I implemented your suggestion I started having keycloak issues that ultimately turned out to be unrelated to this issue, and not the fault of REMS. I will shortly add notes about those issues to the `configuration.md` docs.

Thanks again!
